### PR TITLE
[SPARK-47303][CORE][TESTS] Restructure MasterSuite

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to split `MasterSuite` as below:

```
master
├── MasterSuite.scala
├── MasterSuiteBase.scala
├── MasterWorkerUISuite.scala
├── RecoverySuite.scala
├── ResourceProfilesSuite.scala
└── WorkerSelectionSuite.scala
```

In addition, I reduced the memory and core specification in the tests below:

- Reduced the local cluster's work memory at `WorkerDecommissionExtendedSuite`
- Reduced memory at resource profiles at `ResourceProfilesSuite` and `ResourceProfilesMaxCoresSuite`
- Reduced the number of workers and cores at `ResourceProfilesMaxCoresSuite`

### Why are the changes needed?

MasterSuite takes too much resources, and it makes some tests fail with OOM with Mac OS: https://github.com/apache/spark/actions/runs/8162270810

### Does this PR introduce _any_ user-facing change?

No, test-only.

### How was this patch tested?

https://github.com/HyukjinKwon/spark/actions/runs/8170727017/job/22337438942

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8134128130

**Comment by dongjoon-hyun** [comment]
Thank you for trying this.

File: core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala [hunk 1]
```diff
@@ -731,6 +731,7 @@ class MasterSuite extends SparkFunSuite
       master.registerApplication(appInfo)
       startExecutorsOnWorkers(master)
       assert(appInfo.executors.map(_._2.worker.id).toSeq.distinct.sorted === expected)
+      System.gc()
```
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala:734]: How about adding `System.gc()` in the `after` method.

https://github.com/apache/spark/blob/5a71f218420762bf609087d9c6f7c631fbff1be1/core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala#L163-L169

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8136924661/job/22234172189

**Comment by HyukjinKwon** [comment]
build: https://github.com/HyukjinKwon/spark/actions/runs/8138592236

**Comment by HyukjinKwon** [comment]
test: https://github.com/HyukjinKwon/spark/actions/runs/8149143761/job/22273296949

File: core/src/test/scala/org/apache/spark/deploy/master/WorkerSelectionSuite.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+import org.apache.spark.SparkConf
+import org.apache.spark.deploy.DeployMessages._
+import org.apache.spark.internal.config._
+import org.apache.spark.internal.config.Deploy._
+import org.apache.spark.internal.config.Deploy.WorkerSelectionPolicy._
+import org.apache.spark.internal.config.UI._
+import org.apache.spark.rpc.RpcAddress
+
+class WorkerSelectionSuite extends MasterSuiteBase {
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/deploy/master/WorkerSelectionSuite.scala:30]: +1 for the direction.

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8150215353/job/22276105311

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8151190050

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8152472664/job/22282001033

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8153673605/job/22285532203

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8155204243
https://github.com/HyukjinKwon/spark/actions/runs/8155325532
https://github.com/HyukjinKwon/spark/actions/runs/8155328335

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8164584039

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8165935950/job/22323872479

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8167770830

It should work now I believe but let me wait for the test result.

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8168999430

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8170727017/job/22337438942

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8170741851/job/22337515014

**Comment by HyukjinKwon** [comment]
Once https://github.com/HyukjinKwon/spark/actions/runs/8170727017/job/22337438942 passes, it can be merged.

File: core/src/test/scala/org/apache/spark/deploy/master/WorkerSelectionSuite.scala [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        .set(WORKER_SELECTION_POLICY.key, policy.toString)
+        .set(SPREAD_OUT_APPS.key, spreadOut.toString)
+        .set(UI_ENABLED.key, "false")
+        .set(Network.RPC_NETTY_DISPATCHER_NUM_THREADS, 1)
+        .set(Network.RPC_IO_THREADS, 1)
+      val master = makeAliveMaster(conf)
+
+      // Use different core and memory values to simplify the tests
+      MockWorker.counter.set(10000)
+      (1 to 3).foreach { idx =>
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/deploy/master/WorkerSelectionSuite.scala:56]: BTW, just to be clear, I think the OOM in Mac OS is closely related to the number of workers. It always fails if we create the local worker more than 4.

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/actions/runs/8170727017/job/22337438942 passed.

**Review by LuciferYang — APPROVED** [review]
Nice work~ LGTM ~ Thanks @HyukjinKwon

**Comment by HyukjinKwon** [comment]
Thanks.

Merged to master.

**Review by dongjoon-hyun — COMMENTED** [review]
+1, late LGTM. Thank you so much!

**Pull request closed without merging**